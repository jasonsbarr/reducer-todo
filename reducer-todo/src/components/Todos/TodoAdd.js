import React from "react";

const TodoAdd = {
  return <div>Todo Form</div>
};

export default TodoAdd;
