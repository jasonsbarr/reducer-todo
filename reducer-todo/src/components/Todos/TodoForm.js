import React, { useState } from "react";
import { useLocation } from "react-router-dom";

const defaultValues = {
  item: "",
  due_at_date: "",
  due_at_time: "",
  priority: 4,
};

const TodoForm = ({ handleSubmit, todo = undefined }) => {
  const { pathname } = useLocation();

  const formValues = {
    item: todo ? todo.item : defaultValues.item,
    due_at_date:
      todo && todo.due_at
        ? todo.due_at.toISOString().split("T")[0]
        : defaultValues.due_at_date,
    due_at_time:
      todo && todo.due_at
        ? todo.due_at
            .toISOString()
            .split("T")[1]
            .split(".")[0]
        : defaultValues.due_at_time,
    priority: todo ? todo.priority : defaultValues.priority,
  };

  const [formState, setFormState] = useState(formValues);

  const handleChange = event => {
    setFormState({
      ...formState,
      [event.target.name]: event.target.value,
    });
  };

  return (
    <form onSubmit={handleSubmit}>
      <label htmlFor="item">
        Todo item (required):
        <input
          type="text"
          name="item"
          id="item"
          value={formState.item}
          placeholder="Todo item goes here..."
          onChange={handleChange}
        />
      </label>
      <label htmlFor="due_at">
        Due date:
        <span id="due_at">
          <input
            type="date"
            name="due_at_date"
            id="due_at_date"
            value={formState.due_at_date}
            onChange={handleChange}
          />
          <input
            type="time"
            name="due_at_time"
            id="due_at_time"
            value={formState.due_at_time}
            onChange={handleChange}
          />
        </span>
      </label>
      <label htmlFor="priority">
        Priority:
        <select
          name="priority"
          id="priority"
          value={formState.priority}
          selected={4}
          onChange={handleChange}
        >
          <option value={1}>1</option>
          <option value={2}>2</option>
          <option value={3}>3</option>
          <option value={4}>4</option>
        </select>
      </label>
      <button type="submit">
        {pathname === "/todo/add" ? "Add Todo" : "Edit Todo"}
      </button>
    </form>
  );
};

export default TodoForm;
