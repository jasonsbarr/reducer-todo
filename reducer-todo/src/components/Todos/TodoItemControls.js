import React from "react";
import { useHistory } from "react-router-dom";
import Axios from "axios";
import { DELETE_TODO, EDIT_TODO } from "../../actions/todos";
import useTodos from "../../hooks/useTodos";

const TodoItemControls = ({ todo }) => {
  const apiUrl = `http://localhost:4000/todos`;
  const history = useHistory();
  const [, dispatch] = useTodos();

  const handleCompleteTodo = () => {
    const changed = {
      ...todo,
      completed: !todo.completed,
      updated_at: new Date(Date.now()),
    };

    const options = {
      method: "PUT",
      headers: { "Content-Type": "application/json" },
      data: JSON.stringify(changed),
      url: `${apiUrl}/${todo.id}`,
    };

    dispatch({ type: EDIT_TODO, payload: changed });

    Axios(options).catch(err => console.error(err));
  };

  const handleEditTodo = () => {
    history.push(`/todo/${todo.id}/edit`);
  };

  const handleDeleteTodo = () => {
    dispatch({ type: DELETE_TODO, payload: todo.id });

    const options = {
      method: "DELETE",
      headers: { "Content-Type": "application/json" },
      url: `${apiUrl}/${todo.id}`,
    };

    Axios(options).catch(err => console.error(err));
  };

  return (
    <div className="todo-item-controls">
      <button onClick={handleCompleteTodo}>
        {todo.completed ? "⍻" : "✓"}
      </button>
      <button onClick={handleEditTodo}>🖉</button>
      <button id={`delete-${todo.id}`} onClick={handleDeleteTodo}>
        ×
      </button>
    </div>
  );
};

export default TodoItemControls;
