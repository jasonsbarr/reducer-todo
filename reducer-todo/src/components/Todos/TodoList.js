import React from "react";
import { TodoItem } from ".";

const TodoList = ({ todos }) => {
  return (
    <ul>
      {todos.map(todo => (
        <TodoItem key={todo.id} todo={todo} />
      ))}
      ;
    </ul>
  );
};

export default TodoList;
