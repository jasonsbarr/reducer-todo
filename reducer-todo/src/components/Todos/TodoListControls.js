import React from "react";

const TodoListControls = ({ onClearCompleted }) => (
  <button onClick={onClearCompleted}>Clear Completed</button>
);

export default TodoListControls;
