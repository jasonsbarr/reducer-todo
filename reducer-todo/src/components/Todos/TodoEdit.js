import React from "react";
import { useParams, useHistory } from "react-router-dom";
import useFetchSingleTodo from "../../hooks/useFetchSingleTodo";
import useTodos from "../../hooks/useTodos";
import TodoForm from "./TodoForm";

const TodoEdit = () => {
  const url = `http://localhost:4000/todos`;
  const { todoId } = useParams();
  const [, dispatch] = useTodos();
  const history = useHistory();
  const todo = useFetchSingleTodo(Number(todoId));

  const handleFormSubmit = () => {};

  return (
    <>
      <h1>Edit Todo</h1>
      <div className="form-container">
        <TodoForm todo={todo} handleSubmit={handleFormSubmit} />
      </div>
    </>
  );
};

export default TodoEdit;
