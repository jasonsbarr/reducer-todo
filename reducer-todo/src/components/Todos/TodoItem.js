import React from "react";
import { Link } from "react-router-dom";
import TodoItemControls from "./TodoItemControls";

const TodoItem = ({
  todo,
  onCompleteTodo,
  onEditTodo,
  onDeleteTodo,
} = {}) => {
  const { id, item, priority, due_at, completed, uuid } = todo;

  return (
    <li
      id={`todo-${id}`}
      data-uuid={uuid}
      completed={completed ? 1 : 0}
      className={`todo-item${completed ? " completed" : ""}`}
    >
      <div className="todo-item-container">
        <p>
          <Link to={`/todo/${id}`}>{item}</Link>
        </p>
        <TodoItemControls todo={todo} />
      </div>
    </li>
  );
};

export default TodoItem;
