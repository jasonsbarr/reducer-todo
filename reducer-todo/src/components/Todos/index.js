import TodoItem from "./TodoItem";
import TodoList from "./TodoList";
import TodoAdd from "./TodoAdd";
import TodoSingle from "./TodoSingle";

export { TodoItem, TodoList, TodoAdd, TodoSingle };
