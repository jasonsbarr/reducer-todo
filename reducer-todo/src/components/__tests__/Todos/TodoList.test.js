import React from "react";
import { render, cleanup } from "@testing-library/react";
import { TodoList } from "../../Todos";

afterEach(cleanup);

const todos = [
  {
    item: "Learn about reducers",
    completed: false,
    id: "todo-1",
  },
  {
    item: "Learn more about reducers",
    completed: false,
    id: "todo-2",
  },
  {
    item: "Learn everything there is to know about reducers",
    completed: false,
    id: "todo-3",
  },
];

describe("TodoList component", () => {
  const { container } = render(<TodoList todos={todos} />);
  const todoList = container.firstChild;

  test("should render a TodoList with todos passed in from above", () => {
    expect(todoList.children.length).toEqual(todos.length);
    expect(todoList.textContent).toMatch("Learn about reducers");
    expect(todoList.textContent).toMatch("Learn more about reducers");
    expect(todoList.textContent).toMatch(
      "Learn everything there is to know about reducers",
    );
    expect(todoList).toMatchSnapshot();
  });
});
