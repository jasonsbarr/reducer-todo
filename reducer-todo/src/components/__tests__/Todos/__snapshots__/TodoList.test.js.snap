// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList component should render a TodoList with todos passed in from above 1`] = `
<ul>
  <li
    class="todo-item"
    completed="0"
    id="todo-1"
  >
    Learn about reducers
  </li>
  <li
    class="todo-item"
    completed="0"
    id="todo-2"
  >
    Learn more about reducers
  </li>
  <li
    class="todo-item"
    completed="0"
    id="todo-3"
  >
    Learn everything there is to know about reducers
  </li>
  ;
</ul>
`;
